engine.lisp

```diff
@@ -198,55 +278,82 @@ Returns a vector of matching nodes."
                                (return))))))
         resultset)
     (complete-match-pair (o)
-      (return-from match-pair (value o)))))
+      (return-from match-pair-breadth (value o)))))
 
-(declaim (ftype (function (list (or plump:node vector list))
+(declaim (ftype (function (list (or plump:node vector list) &optional keyword)
                           (values (and (vector plump:node) (not simple-array))))
                 match-selector))
-(defun match-group (group root-node)
+(defun match-group (group root-node &optional (search-type :DFS))
   "Match a matcher group against the root-node and possibly all its children.
 Returns an array of mached nodes."
   (declare (optimize speed))
   (assert (eq (car group) :group) () 'selector-malformed)
   (let ((group (cdr group)))
-    (loop with nodes = (etypecase root-node
-                         (plump:node (make-array 1 :initial-element root-node :adjustable T :fill-pointer T))
-                         (vector root-node)
-                         (list (coerce root-node 'vector)))
-          for combinator = (pop group)
-          for matcher = (pop group)
-          while matcher
-          do (setf nodes (match-pair combinator matcher nodes))
-          finally (return nodes))))
+    (ecase search-type
+        (:DFS
+         (let* ((result (make-array 10 :adjustable T :fill-pointer 0)))
+           (labels ((add-to-result (node)
+                      (vector-push-extend node result))
+                    (search-node (node group)
+                      (let ((combinator (car group))
+                            (matcher (cadr group))
+                            (group (cddr group)))
+                        (if (second group)
```

Comment:
No we can't as this means we recurse one more time which is bad. It could be replaced by `group` I guess.